### PR TITLE
[DOCS] Cross-reference API/deep dives/tutorials

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

As I've been combing through our docs to write a recipe landing page, I've found we sometimes link directly to our GitHub code when there's an API doc page that exists. Our API docs are 🤌🏽. Our deep dives and tutorials are 🤌🏽. Let's make sure they're properly cross-referenced 🤌🏽🤌🏽 so our users don't have to go hunting for any details when reading.

#### Test plan

```
👁️                          👁️
        __
```

### CODE REVIEW

**Review by joecummings — COMMENTED** [review]
One comment. The rest is 🔥

File: docs/source/api_ref_models.rst [hunk 1]
```diff
@@ -237,3 +237,22 @@ Vision components to support multimodality using `CLIP encoder <https://arxiv.or
     clip.TokenPositionalEmbedding
     clip.TiledTokenPositionalEmbedding
     clip.TilePositionalEmbedding
+
+
+.. _convert_weights_label:
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/api_ref_models.rst:242]: Yeah it looks particularly weird when you see `convert_weights` in the `torchtune.models` docs. I can see why we want to contain per-model `convert_weight` utils in each model's module, so this kind-of follows that format.

Will users particularly care for this detail of how we convert convert weights? Couldn't we just omit the code reference and also the `convert_weights` api docs? 

If we do want to ref code, I would go for something like [this](https://github.com/pytorch/torchtune/blob/e10142016798cf84f2e5c638a985014384f400a7/torchtune/utils/_checkpointing/_checkpointer.py#L348) and [this](https://github.com/pytorch/torchtune/blob/e10142016798cf84f2e5c638a985014384f400a7/torchtune/utils/_checkpointing/_checkpointer.py#L436) instead.

[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/api_ref_models.rst:242]: I've simplified things by trying to explain things better in the deepdive without explicitly exposing our weight conversion utilities.

File: docs/source/tutorials/lora_finetune.rst [hunk 2]
```diff
@@ -315,13 +318,13 @@ to see our peak memory during a finetune. We will experiment along two axes:
...
 
-To compare the results of our experiments, we can evaluate our models on :code:`truthfulqa_mc2`, a task from
+To compare the results of our experiments, we can evaluate our models on `truthfulqa_mc2 <https://github.com/sylinrl/TruthfulQA>`_, a task from
 the `TruthfulQA <https://arxiv.org/abs/2109.07958>`_ benchmark for language models. For more details on how to run this and other evaluation tasks
 with torchtune's EleutherAI evaluation harness integration, see our :ref:`End-to-End Workflow Tutorial <eval_harness_label>`.
 
 Previously, we only enabled LoRA for the linear layers in each self-attention module, but in fact there are other linear
 layers we can apply LoRA to: MLP layers and our model's final output projection. Note that for Llama-2-7B the final output
-projection maps to dimension 32000 (instead of 4096 as in the other linear layers), so enabling LoRA for this layer will increase
+projection maps to the vocability dimension (32000 instead of 4096 as in the other linear layers), so enabling LoRA for this layer will increase
```
Comment (felipemello1) [inline docs/source/tutorials/lora_finetune.rst:327]: just confirming (english is not 1st language), is vocabilty the right word? Should it be vocabulary?

**Review by felipemello1 — COMMENTED** [review]
its nice to have this refresh on all of the links/references. I saw that you also added some for wikis, like bf16.

File: docs/source/tutorials/lora_finetune.rst (hunk 2 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/lora_finetune.rst:327]: your English is perfect Felipe! it's a good catch

File: docs/source/deep_dives/checkpointer.rst [hunk 3]
```diff
@@ -115,21 +115,28 @@ fine-tuned checkpoints from torchtune with any post-training tool (quantization,
...
-To be "state-dict invariant", the ``load_checkpoint`` and
-``save_checkpoint`` methods make use of the weight convertors available
-`here <https://github.com/pytorch/torchtune/blob/main/torchtune/models/convert_weights.py>`_.
+.. note::
+
+  To be state-dict "invariant" in this way, the ``load_checkpoint`` and ``save_checkpoint`` methods of each checkpointer
+  make use of weight convertors which correctly map weights between checkpoint formats. For example, when loading weights
+  from HuggingFace, we apply a permutation to certain weights on load and save to ensure checkpoints behave exactly the same.
+  To further illustrate this, the llama family of models uses a
+  `generic weight converter function <https://github.com/pytorch/torchtune/blob/898670f0eb58f956b5228e5a55ccac4ea0efaff8/torchtune/models/convert_weights.py#L113>`_
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/deep_dives/checkpointer.rst:124]: Yeah, ideally we expose this correctly through the docs. However, we don't generate docs for anything inside `convert_weights`. As mentioned [above](https://github.com/pytorch/torchtune/pull/1229#discussion_r1693148724), exposing `hf_to_tune` etc. in the `modules` api ref looks really weird, and, as Joe mentioned, we want to contextualise these functions to only be used inside of checkpointers - we don't want users thinking they have to convert weights using these functions.

I think I'd like to see another issue addressing where our model conversion logic lives so our api doctree looks sane.

**Comment by felipemello1** [comment]
LGTM, up to @joecummings

File: docs/source/deep_dives/checkpointer.rst [hunk 4]
```diff
@@ -115,21 +115,28 @@ fine-tuned checkpoints from torchtune with any post-training tool (quantization,
 which supports the source format, without any code changes OR conversion scripts. This is one of the
 ways in which torchtune interoperates with the surrounding ecosystem.
 
-To be "state-dict invariant", the ``load_checkpoint`` and
-``save_checkpoint`` methods make use of the weight convertors available
-`here <https://github.com/pytorch/torchtune/blob/main/torchtune/models/convert_weights.py>`_.
+.. note::
+
+  To be state-dict "invariant" in this way, the ``load_checkpoint`` and ``save_checkpoint`` methods of each checkpointer
+  make use of weight convertors which correctly map weights between checkpoint formats. For example, when loading weights
```
Comment (RdoubleA) [inline docs/source/deep_dives/checkpointer.rst:121]:
```suggestion
  make use of weight converters which correctly map weights between checkpoint formats. For example, when loading weights
```

File: docs/source/deep_dives/checkpointer.rst [hunk 5]
```diff
@@ -115,21 +115,28 @@ fine-tuned checkpoints from torchtune with any post-training tool (quantization,
...
+  from HuggingFace, we apply a permutation to certain weights on load and save to ensure checkpoints behave exactly the same.
```
Comment (RdoubleA) [inline docs/source/deep_dives/checkpointer.rst:122]:
```suggestion
  from Hugging Face, we apply a permutation to certain weights on load and save to ensure checkpoints behave exactly the same.
```
is permutation the right word? maybe "remap certain weights"?

File: docs/source/deep_dives/checkpointer.rst [hunk 6]
```diff
@@ -115,21 +115,28 @@ fine-tuned checkpoints from torchtune with any post-training tool (quantization,
...
+  To further illustrate this, the llama family of models uses a
```
Comment (RdoubleA) [inline docs/source/deep_dives/checkpointer.rst:123]:
```suggestion
  To further illustrate this, the Llama family of models uses a
```

File: docs/source/deep_dives/checkpointer.rst [hunk 7]
```diff
@@ -115,21 +115,28 @@ fine-tuned checkpoints from torchtune with any post-training tool (quantization,
...
+  `generic weight converter function <https://github.com/pytorch/torchtune/blob/898670f0eb58f956b5228e5a55ccac4ea0efaff8/torchtune/models/convert_weights.py#L113>`_
+  whilst some other models like Phi have their own `conversion functions <https://github.com/pytorch/torchtune/blob/main/torchtune/models/phi3/_convert_weights.py>`_
```
Comment (RdoubleA) [inline docs/source/deep_dives/checkpointer.rst:125]:
```suggestion
  whilst some other models like Phi3 have their own `conversion functions <https://github.com/pytorch/torchtune/blob/main/torchtune/models/phi3/_convert_weights.py>`_
```

File: docs/source/tutorials/llama3.rst [hunk 8]
```diff
@@ -78,15 +78,15 @@ We can also add command-line overrides as needed, e.g.
...
 
 .. note::
-    To see the full set of configurable parameters for this (and other) configs we can use ``tune cp`` to copy (and modify)
-    the default config. ``tune cp`` can be used with recipe scripts too, in case you want to make more custom changes
-    that cannot be achieved by directly modifying existing configurable parameters. For more on ``tune cp`` see the section on
-    :ref:`modifying configs <tune_cp_label>`.
+    To see the full set of configurable parameters for this (and other) configs we can use :ref:`tune cp <tune_cp_cli_label>` to copy (and modify)
+    the default config. :ref:`tune cp <tune_cp_cli_label>` can be used with recipe scripts too, in case you want to make more custom changes
+    that cannot be achieved by directly modifying existing configurable parameters. For more on :ref:`tune cp <tune_cp_cli_label>` see the section on
+    :ref:`modifying configs <tune_cp_label>` in our ":ref:`finetune_llama_label`" tutorial.
```
Comment (RdoubleA) [inline docs/source/tutorials/llama3.rst:89]: need <> here?

File: docs/source/tutorials/qat_finetune.rst [hunk 9]
```diff
@@ -106,7 +106,7 @@ is ready for fine-tuning.
 
 After fine-tuning, we can convert the model to get an actual quantized model.
 If we print the converted model, we’ll see that the QAT linears have been
-swapped with :code:`Int8DynActInt4WeightLinear`, which are the quantized versions
+swapped with `Int8DynActInt4WeightLinear <https://github.com/pytorch/ao/blob/428084356ace4ea94c22a3a9b3d74cff8ee41db3/torchao/quantization/prototype/qat.py#L38>`_, which are the quantized versions
```
Comment (RdoubleA) [inline docs/source/tutorials/qat_finetune.rst:109]: does torchao have an API ref instead?

File: docs/source/tutorials/qlora_finetune.rst [hunk 10]
```diff
@@ -31,10 +31,10 @@ What is QLoRA?
 QLoRA builds on top of LoRA to enable further
 memory savings. In LoRA, model parameters can be thought of as existing in two partitions: adapters, which are
 low-rank matrices added to different layers of a neural network, and base model parameters, which are parameters that are part of
-the original model. In vanilla LoRA-style training, both these parameters are held in the same precision (typically fp32 or bf16), and
+the original model. In vanilla LoRA-style training, both these parameters are held in the same precision (typically fp32 or `bf16 <https://en.wikipedia.org/wiki/Bfloat16_floating-point_format#bfloat16_floating-point_format>`_.), and
 therefore activations and intermediate gradients computed are in fp32/bf16.
 
-QLoRA further quantizes the base model parameters into a bespoke 4-bit NormalFloat (NF4) data type, resulting in 4-8x less parameter memory usage while
+QLoRA further quantizes the base model parameters into a bespoke 4-bit NormalFloat (`NF4 <https://www.youtube.com/watch?v=TPcXVJ1VSRI&t=563s>`_) data type, resulting in 4-8x less parameter memory usage while
```
Comment (RdoubleA) [inline docs/source/tutorials/qlora_finetune.rst:37]: was half expecting this youtube link to be a rick roll

**Review by RdoubleA — APPROVED** [review]
some nit comments, otherwise I'm onboard with these changes.

File: docs/source/tutorials/qlora_finetune.rst (hunk 10 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/qlora_finetune.rst:37]: I would never do that to you. I actually wonder if we should support the official PyTorch talk [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) instead?

Comment (RdoubleA) [reply docs/source/tutorials/qlora_finetune.rst:37]: I knew that was coming and I still clicked on it

File: docs/source/tutorials/llama3.rst (hunk 8 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/llama3.rst:89]: just grabs the title if no text provided

**Pull request merged**